### DISCUSSION TITLE
Using test security with dynamic users

### DISCUSSION THREAD

**Discussion by eijckron** [quarkusio/quarkus#42309] [Q&A]
Quarkus test security allows to run a test with a specified user and roles.
For example these annotations define a test that is executed as user testUser with a JWT that has sub set to test-user-id.
```Java
@Test
@TestSecurity(user = "testUser")
@OidcSecurity(claims = @Claim(key = "sub", value = "test-user-id"))
testMethod()
```

However I have a couple of use cases where I would like to define the security for the call dynamically like:
```Java
@Test
testMethod() {
    given().auth().oAuth2(mockATestUser("testUser", "test-user-id", "ROLE-A")
}
```
Besides setting up keycloak dev services with all the possible dynamic users I have not found any references to a setup that would allow to have these test mock users behave in the same way as the test mock users defined through the annotations.

Scenario's that would benefit from this dynamic behavior:
* Parameterized tests run for different role configurations
* Nested tests where each invocation needs its own user context to avoid clashes between test cases.

[a comment by an automated account is not shown]

**Comment by sberyozkin** [comment]
@eijckron Hi, do each of these new user/role combinations touch on different endpoint code paths ?

**Reply by eijckron** [reply]
@sberyozkin 
The test is usually for a specific endpoint but could potentially also be a variable.
A typical use case is endpoint access validation. A parameterized test provides a set of roles that should / should not have access.
But it could go as far as a single parameterized test that receives endpoint, role and expected access true/false
